service/src/main.rs

```diff
@@ -527,55 +539,73 @@ pub fn init_light_client<E: EnclaveBase + SideChain>(
 		.init_light_client(genesis_header, authority_list, grandpa_proof)
 		.unwrap();
 
-	info!("Finished initializing light client, syncing....");
+	info!("Finished initializing light client, syncing parent chain...");
+
+	let latest_synced_header = sync_parentchain(enclave_api, api, &latest);
+
+	info!("Execute trusted operations for the first time and start side chain block production");
+
+	execute_trusted_operations(enclave_api);
+
+	latest_synced_header
+}
+
+/// Subscribe to the node API finalized heads stream and trigger a parent chain sync
+/// upon receiving a new header
+fn subscribe_to_parentchain_new_headers<E: EnclaveBase + SideChain>(
+	enclave_api: &E,
+	api: &Api<sr25519::Pair, WsRpcClient>,
+	mut last_synced_header: Header,
+) -> Result<(), Error> {
+	let (sender, receiver) = channel();
+	api.subscribe_finalized_heads(sender).map_err(Error::ApiClientError)?;
+
+	loop {
+		let new_header: Header = match receiver.recv() {
+			Ok(header_str) => serde_json::from_str(&header_str).map_err(Error::Serialization),
+			Err(e) => Err(Error::ApiSubscriptionDisconnected(e)),
+		}?;
+
+		println!(
+			"[+] Received finalized header update ({}), syncing parent chain...",
+			new_header.number
+		);
 
-	sync_parentchain_and_execute_tops(enclave_api, api, latest)
+		last_synced_header = sync_parentchain(enclave_api, api, &last_synced_header);
+	}
 }
 
 /// Gets the amount of blocks to sync from the parentchain and feeds them to the enclave.
 ///
-/// It also calls into the enclave if there are no blocks to sync because the enclave does
-/// currently two things in `produce_blocks`:
-///
-/// * sync blocks
-/// * execute pending trusted operations
 ///
-/// Todo: the two tasks above should be independent: #404
-pub fn sync_parentchain_and_execute_tops<E: EnclaveBase + SideChain>(
+pub fn sync_parentchain<E: EnclaveBase + SideChain>(
 	enclave_api: &E,
 	api: &Api<sr25519::Pair, WsRpcClient>,
-	mut last_synced_header: Header,
+	last_synced_header: &Header,
```

Comment:
This change requires you to use the additional clone below. What was the reason for this change?

In this case, it might be actually relevant as thousands of headers are synced.